### PR TITLE
Large-scale structure simulation code

### PR DESCRIPTION
**Pull request by jrs65**

### CODE REVIEW

**Comment by jrs65** [comment]
@sjforeman I think it's probably time we get this thing merged.

I'm trying to think what the major changes since you last looked are...

- Massive improvement to the correlation function calculations. More accurate and way way faster (Richardson extrapolation kicks ass!). It's so fast now you basically don't need to pregenerate the correlation functions any more.
- Parallelisation of the C_l(z, z') array calculation. This is a speed up for ~4x relative to the old version.

File: cora/signal/lss.py [hunk 1]
```diff
@@ -0,0 +1,933 @@
...
+            fr = cosmology.growth_rate(za, c)
+            vpsi[0] *= (1 + fr)[:, np.newaxis]
+
+        # Create the final field container
+        final_field = BiasedLSS(axes_from=biased_field, attrs_from=biased_field)
+        final_field.delta[:] = 0.0
+
+        # Calculate the Zel'dovich displaced field and assign directly into the output
+        # container
+        # za_density(vpsi, biased_field.delta[:], nside, chi, out=final_field.delta[:])
```
Comment (sjforeman) [inline cora/signal/lss.py:524]: Delete commented out code

File: cora/signal/lss.py [hunk 2]
```diff
@@ -0,0 +1,933 @@
...
+    chi_ext = np.zeros(len(chi) + 2, dtype=chi.dtype)
+    chi_ext[1:-1] = chi
+    chi_ext[0] = chi[0] - (chi[1] - chi[0])
+    chi_ext[-1] = chi[-1] + (chi[-1] - chi[-2])
+
+    # Array to hold the density obtained from the ZA approximation
+    if out is None:
+        out = np.zeros((nchi, npix), dtype=delta_bias.dtype)
+
+    angpos = np.array(healpy.pix2ang(nside, np.arange(npix)))
```
Comment (sjforeman) [inline cora/signal/lss.py:720]: Duplicate of line 707

File: cora/signal/lss.py [hunk 3]
```diff
@@ -0,0 +1,933 @@
...
+        # Final angles and radial coordinate
+        new_angpos = calculate_positions(angpos, psi_slc[1:])
+        new_chi = chi[ii] + psi_slc[0]
+
+        # Get the weights for the surrounding pixels
+        pixel_ind, pixel_weight = healpy.get_interp_weights(
+            nside, new_angpos[0], new_angpos[1]
+        )
+
+        # Find
```
Comment (sjforeman) [inline cora/signal/lss.py:738]: Find...?

File: cora/util/pmesh.pyx [hunk 4]
```diff
@@ -0,0 +1,452 @@
+# cython: profile=True
+
+""" Module for deriving a density filed from a distribution of particles. """
```
Comment (sjforeman) [inline cora/util/pmesh.pyx:3]: Typo: filed -> field

File: cora/util/pmesh.pyx [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+# degraded the performance slightly.
+cpdef void fill_delta_za(double[:,:] delta_za, long[:] rad_idx,
+                         long[:] ang_idx, double[:] rho0):
+# 'delta_za' is defined in the argument as a cython memoryview.
+# It can take a numpy array as parameter and modifying one changes the other.
+    cdef long ii, jj, kk
+    for kk in range(len(rad_idx)):
+        ii = rad_idx[kk]
+        jj = ang_idx[kk]
+        #delta_za[ii,jj] += 1. # TODO: delete
```
Comment (sjforeman) [inline cora/util/pmesh.pyx:129]: Delete old code

**Review by sjforeman — COMMENTED** [review]
This looks good overall - my comments are mostly very minor. However, I'm still slightly concerned about the disagreement between linear+RSD maps produced by this code and the flat-sky angular power spectrum computed by the existing code.

Here are a few comparisons, evaluated at a frequency in the middle of the band and at the lower edge of the band.
midband
loweredge
A mismatch is expected at the band edges, due to the poor performance of the one-sided second derivatives there. However, I'm also finding a 10% mismatch at ell=500, which grows to 20% at ell=1500. This mismatch might make comparisons of simulations and predictions more difficult, unless we measure or derive a correction factor that accounts for the difference.

For stacking, I think we'll want to quantify this by comparing sims with this new code and with a flat-sky input angular power spectrum. More generally, though, we might want to at least add some comments about the mismatch to a relevant docstring...

**Comment by jrs65** [comment]
I don't find a difference quite as dramatic, although it's still not ideal...
Unknown-48

More like 5% at l=500, and 10% at l=1500. That was with xromb=3 set for both the original cora code and for the new code.

**Comment by sjforeman** [comment]
Yeah, I also did a test with `xromb=3`, and found similar results to you:
midband_xromb3

**Comment by jrs65** [comment]
Yeah, xromb seems to make a lot of difference:
Unknown-49

**Comment by jrs65** [comment]
Oh. Just seen you commented about the same thing.

Overall, I'm not that concerned about xromb=0. If you don't smooth over the channel width you are going to be sensitive to extremely small scales, and with the different ways that the power spectrum is regularized for each scheme and the FD derivatives you would never expect that to work very well. But it would be nice to get things a bit closer for the xromb > 0 case.

To give a better synthesis of which scales are important, I smoothed each map by a FWHM=0.25 degree Gaussian (which is comparable to CHIME's resolution). The new code produces maps that have an RMS 1.8% lower.

**Comment by sjforeman** [comment]
I see your points about xromb=0 not being expected to work as well. I wonder what would happen if some sort of radial smoothing is applied before taking derivatives of the maps? Most of the power in LSS velocities is at large scales, so smoothing shouldn't result in much physical information loss, but might help with the numerics of the radial derivatives... We could also experiment with oversampling in the radial direction before taking derivatives, and then downsampling afterwards.

Regardless, it's probably fine to merge this now and leave further refinements as a to-do list item. However, let's leave a comment in a docstring somewhere, in case a future use finds a high-ell discrepancy and pulls their hair out trying to fix it.

**Comment by jrs65** [comment]
Okay. Here's a bit of fun progress.

First, one thing this isn't. There's an interesting term that comes from the ordering of the channel smoothing and the second derivative of the potential, in this new scheme they are applied in the order above, but physically it's the other way around. Now if the channel width was constant in comoving distance they would commute, but as they aren't there is a correction term which is roughly the first derivative of the potential * first derivative of channel widths. This turns out to be small for large l, but is order unit for the largest scales (l < 5).

So back to the finite difference derivative scheme. As we've seen before the issue does seem to be in the finite difference scheme, simply upsampling the redshift direction by a factor of two, and then averaging gives significantly improved results, with the obvious problem that this increases the computation/memory required to generate it. 

However, since I've recently fallen in love with Richardson extrapolation I thought I'd give it a go in this case, and it's a significant improvement for barely any extra work:
Unknown-50
The error is roughly halved at the largest l's, and the beam smooth RMS residuals go down from 1.8% to 0.6%.

I've just done one order of Richardson extrapolation in here. The main thing you need is to have the second derivative at twice the step size, which can be done by partitioning the field into even/odd frequency channels, taking the FD second derivative of each, and then reassembling into a single field which essentially used twice the step size (although the changing channel width breaks this slightly). Then you just do a Richardson step: d2_richardson = (4 * d2_highres - d2_lowres) / 3

In theory you can do more extrapolation orders to improve the accuracy for a roughly linear computational cost (i.e. needing to calculate the derivatives multiple times), however you either need to keep many copies of the field in memory, or build it into the `lssutil.diff2` routine and do a lot of book keeping.

A final caveat is that I think this increases the edge effects as you essentially ended up in the one sided regime further away from the edge than before.

**Comment by jrs65** [comment]
Actually, I think the first point above (the varying channel width issues) actually isn't a problem as the code uses a fixed channel width in comoving distance for smoothing at the moment. If we ever correct that we might care about it.

File: cora/signal/corrfunc.py [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+        )
+    else:
+        rhigh, Fhigh = _corr_hankel(
+            psfunc, switchlogr, maxlogr, samples_per_decade, **kwargs
+        )
+
+    # Explicitly calculate and insert the zero lag correlation
+    # TODO: remove the hardcoded limits
+    rlow = np.insert(rlow, 0, 0.0)
+    Flow = _corr_direct(psfunc, -5, 3, rlow)
```
[the remark this answers is not in the document]
Comment (jrs65) [reply cora/signal/corrfunc.py:255]: I don't think so. Those are just the minimum and maximum log-k's being used. I think in the end I'm happy hardcoding the limits, but I'll modify the docstring to say what they are.

[the remark this answers is not in the document]
Comment (jrs65) [reply cora/signal/corrfunc.py:255]: Actually scratch that, I'll make them optional parameters.

File: cora/signal/lss.py [hunk 7]
```diff
@@ -0,0 +1,933 @@
...
+    Attributes
+    ----------
+    nside : int
+        The Healpix resolution to use.
+    redshift : np.ndarray
+        The redshifts for the map slices.
+    frequencies : np.ndarray
+        The frequencies that determine the redshifts for the map slices.
+        Overrides redshift property if specified.
+    num : int
```
[the remark this answers is not in the document]
Comment (jrs65) [reply cora/signal/lss.py:151]: Agreed.

File: cora/signal/lss.py [hunk 8]
```diff
@@ -0,0 +1,933 @@
...
+class GenerateInitialLSS(task.SingleTask):
+    """Generate the initial LSS distribution.
+
+    Attributes
+    ----------
+    nside : int
+        The Healpix resolution to use.
+    redshift : np.ndarray
+        The redshifts for the map slices.
+    frequencies : np.ndarray
```
[the remark this answers is not in the document]
Comment (jrs65) [reply cora/signal/lss.py:148]: Yeah, I agree this is very annoying. There's not a great way to do it at the moment, but I'll create an issue for it.

[the remark this answers is not in the document]
Comment (jrs65) [reply cora/signal/lss.py:148]: See #48

File: cora/signal/lss.py [hunk 9]
```diff
@@ -0,0 +1,933 @@
...
+            biased_field.delta[:] += (D * b1)[
+                loff : loff + lshape, np.newaxis
+            ] * f.delta[:]
+        except NotImplementedError:
+            self.log.info("First order bias is not implemented. This is a bit odd.")
+
+        # Apply any second order bias
+        try:
+            b2 = self._bias_2(z)
+            # NOTE: should this average be over redshift shells, not the full volume?
```
[the remark this answers is not in the document]
Comment (jrs65) [reply cora/signal/lss.py:354]: I agree. Changed!

File: cora/signal/lss.py [hunk 10]
```diff
@@ -0,0 +1,933 @@
...
+    """Generate a field with a constant linear bias.
+
+    Attributes
+    ----------
+    bias : float
+        The Lagrangian bias. At first order this is the related to the Eulerian bias
+        by :math:`b_L = b_E - 1`.
+    """
+
+    bias = config.Property(proptype=float, default=0.0)
```
[the remark this answers is not in the document]
Comment (jrs65) [reply cora/signal/lss.py:390]: Agreed. Changed to `bias_L`

**Comment by jrs65** [comment]
Okay. Changes made!

**Pull request merged**